lib/anagrams.rb

```diff
@@ -1,2 +1,11 @@
 def anagrams?(str1, str2)
+    a = []
+    b = []
+        a<<str1.downcase.chars.sort.join
+        b<<str2.downcase.chars.sort.join
+end
```

Comment:
The error you're seeing it caused by this `end` (here on line 6).

That `end` is ending the method itself. Here's your code again, indented properly:

```ruby
def anagrams?(str1, str2)
    a = []
    b = []
    a<<str1.downcase.chars.sort.join
    b<<str2.downcase.chars.sort.join
end

if a == b
    puts "true"
else
    puts "false"
end
```

Looking at it now, you can see visually that the `anagrams?` method is defined in the first 6 lines of code. After that–after the `end`–the code is no longer part of the method.

That's why running the code results in an error complaining that `a` is undefined. Because by the 8th line of the code (`if a==b`) the variable `a` is not defined. It's only defined within the context of the method definition.

---

If you're able to figure out where to move that `end`, you'll have one other correction to make:

Your method should not end with `puts "true"`. The `puts` command is used to display text to someone, but a _method_ is a more pure operation. That is, a method should just arrive at some result based on an algorithm defined within. If you write `puts theResult`, then the method is actually doing two things: 1) It's arriving at the result, and 2) It's displaying the result.

_Metaphor Alert:_

If we're throwing a party and I ask you how many bottles of wine to buy, you'll execute an algorithm in your mind to figure out the answer. E.g. you'll count up people, subtract those who are kids and don't like to drink, add a couple bottle to err on the side of having enough, etc. Then you'll have the `result`.

Notice that the algorithm **does not** include "Send a text to Sumeet". Because if it did, and I was standing right in front of you when I asked you the question, it would be weird that you sent me a text message with the amount. Conversely, the algorithm doesn't include "Tell Sumeet to his face", because if I was at the store trying to figure out how many bottles to buy, it would be annoying that you refused to tell me until I was face-to-face with you.

The algorithm merely lists the steps needed to achieve the result. The end of the algorithm is `return whateverTheResultIs`. _Other_ parts of our code might use the method in conjunction with `puts` to display the answer to someone. But the algorithm should be free of `puts` to avoid marrying the _algorithm's result_ with the _medium it's expressed_.

So your method should end with `return true`. (Notice that's not `return "true"` – because `true` without quotes is an actual `true/false` value (known as a "boolean"), whereas `"true"` with quotes is just a string of characters.)